### ISSUE TITLE
Respecting `.cargo/config.toml`

### ISSUE THREAD

**Issue by Veykril** [rust-analyzer/rust-analyzer#8741]
This is more of a collection of some issues floating around regarding the `.cargo/config.toml` and general discussion of how to solve these.

- ~~Respect the `build.target` key for analysis(can be worked around mostly with `"rust-analyzer.cargo.target"`): https://github.com/rust-analyzer/rust-analyzer/issues/6604 https://github.com/rust-analyzer/rust-analyzer/issues/5904 https://github.com/rust-analyzer/rust-analyzer/issues/8430~~
- ~~Respect features set via rustflags for analysis https://github.com/rust-analyzer/rust-analyzer/issues/8480~~
- rustflags causes rebuilds https://github.com/rust-analyzer/rust-analyzer/issues/5828

As pointed out in https://github.com/rust-analyzer/rust-analyzer/issues/5904#issuecomment-749187860:
> I don't see any reason right now why `.cargo/config` wouldn't already have an effect -- we're just calling `cargo check`. Or do you mean also using the target from `.cargo/config` for analysis (i.e. `rust-analyzer.cargo.target`)? Maybe `cargo metadata` doesn't use that configuration, or we call it in a way that prevents that; that might be worth investigating. We probably don't want to parse `.cargo/config` ourselves though.

We probably do not want to parse this ourselves but `cargo metadata` doesn't seem to expose this info in its output either, see https://doc.rust-lang.org/cargo/commands/cargo-metadata.html#output-format. So we either have to parse it ourselves or get the relevant info to be emitted by `cargo metadata` from what I can tell.

**Comment by flodiebold** [comment]
Note that we explicitly pass `--filter-platform` to `cargo metadata`, so I'm not sure whether it would even be possible for it to respect the target config.

**Comment by Veykril** [comment]
Hm, ye that might be a problem. I wonder if parsing the config might be not such a bad idea though, we wouldn't have to parse everything but a few things we are interested in anyways(currently `build.rustflags` and `build.target`)?
Though I guess it becomes a bit more complex when it comes to `target.<triple>.rustflags`.

**Comment by Veykril** [comment]
Then again we don't have toml in our dependency tree and I don't think we would want to add that. I wonder if it would make sense to add a subcommand to cargo to emit the configuration as json similar to what `metadata` does for packages? That way cargo would also be the one to resolve the paths to the configs not requiring us to search for them in the appropriate folders.

**Comment by Veykril** [comment]
Actually, I think we can get by with this https://github.com/rust-lang/cargo/issues/9357 once it lands?

**Comment by Veykril** [comment]
Managed to through a working branch together: https://github.com/Veykril/rust-analyzer/commit/8989fb8315538aece975663c3be4aba867e9ee86

![f1Gup1aiAn](https://user-images.githubusercontent.com/3757771/117545448-1dcaae00-b026-11eb-977a-0f35a5e3f2e0.gif)

This uses unstable cargo features, `cargo config get` to read the target triple out of the config to pass to `cargo metadata` --filter-platform and `cargo rustc --print` to read out the `rustc_cfgs`. If those commands fail they fall back to what we previously used, namely invoking rustc directly.

Now the question is, as we don't really use nightly apis, does not using unstable things also apply to cargo when we have fallbacks in place if those commands fail?

Anyways at least we got a (future) solution to these problems.
